### PR TITLE
Refactor GeoGridTiler classes

### PR DESCRIPTION
**Pull request by iverase**
While working on improving how bounded geogrid aggregations work, I realise that the current hierarchy of classes was confusing and difficult to work with. ThisPR modifies this hierarchy by adding an `AbstractGeoHashGridTiler` and `AbstractGeoTileGridTiler`.

### CODE REVIEW

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoTileGridTiler.java [hunk 1]
```diff
@@ -33,69 +43,60 @@ public int advancePointValue(long[] values, double x, double y, int precision, i
...
 
     /**
      * Sets a singular doc-value for the {@link GeoShapeValues.GeoShapeValue}. To be overriden by {@link BoundedGeoTileGridTiler}
      * to account for {@link org.elasticsearch.common.geo.GeoBoundingBox} conditions
      */
-    protected int setValue(GeoShapeCellValues docValues, GeoShapeValues.GeoShapeValue geoValue, int xTile, int yTile, int precision) {
-        docValues.resizeCell(1);
-        docValues.add(0, GeoTileUtils.longEncodeTiles(precision, xTile, yTile));
-        return 1;
+    protected int setValue(GeoShapeCellValues docValues, GeoShapeValues.GeoShapeValue geoValue, int xTile, int yTile) {
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoTileGridTiler.java:83]: geoValue doesn't seemed to be used and the java docs looks seems to be obsolete

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java [hunk 2]
```diff
@@ -172,11 +172,11 @@ private void registerGeoShapeGridAggregators(ValuesSourceRegistry.Builder builde
                 if (getLicenseState().checkFeature(XPackLicenseState.Feature.SPATIAL_GEO_GRID)) {
                     final GeoGridTiler tiler;
                     if (geoBoundingBox.isUnbounded()) {
-                        tiler = new GeoTileGridTiler();
+                        tiler = new UnBoundedGeoTileGridTiler(precision);
```
Comment (imotov) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java:175]: I have struggling a bit over this one. There are two ways to approach this - make tiler a singleton and pass everything as parameters to calls or pass or agg-level parameter here. I am fine with ether, the mix was a bit weird, so I think it is a good change.

**Review by imotov — APPROVED** [review]
LGTM

**Comment by iverase** [comment]
@elasticmachine update branch

**Pull request merged**